lms/views/decorators/h_api.py

```diff
@@ -122,3 +117,122 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
+    #   decorator.
+    #
+    # - The views that this decorator decorates expect a ``jwt`` argument so this
+    #   decorator has to pass it to them (or rather it has to pass it down to the
+    #   next decorator down in the stack, and it eventually gets passed to the
+    #   view).
+    #
+    # This should all be refactored so that views and view decorators aren't
+    # tightly coupled and arguments don't need to be passed through multiple
+    # decorators to the view.
+    def wrapper(request, jwt):
+        _maybe_create_group(request)
+        return wrapped(request, jwt)
+    return wrapper
+
+
+def _maybe_create_group(request):
+    """Create a Hypothesis group for the LTI course, if one doesn't exist."""
+    # Only create groups for application instances that we've enabled the
+    # auto provisioning features for.
+    if not _auto_provisioning_feature_enabled(request):
+        return
+
+    get_param = functools.partial(_get_param, request)
+
+    # Show the user an error page if the group hasn't been created yet and
+    # the user isn't allowed to create groups.
+    if not any(role in get_param("roles").lower() for role in ("administrator", "instructor", "teachingassisstant")):
+        raise HTTPBadRequest("Instructor must launch assignment first")
+
+    tool_consumer_instance_guid = get_param("tool_consumer_instance_guid")
+    context_id = get_param("context_id")
+
+    group = models.CourseGroup.get(request.db, tool_consumer_instance_guid, context_id)
+
+    if group:
+        # The group has already been created in h, nothing more to do here.
+        return
+
+    create_group_api_url = request.registry.settings["h_api_url"] + "/groups"
+    client_id = request.registry.settings["h_client_id"]
+    client_secret = request.registry.settings["h_client_secret"]
+
+    # Generate the name for the new group.
+    try:
+        name = util.generate_group_name(request.params)
+    except MissingContextTitleError:
+        raise HTTPBadRequest('Required parameter "context_title" missing from LTI params')
+
+    authority = request.registry.settings["h_authority"]
+    username = util.generate_username(request.params)
+
+    # Create the group in h.
+    try:
+        response = requests.post(
+            create_group_api_url,
+            auth=(client_id, client_secret),
+            data=json.dumps({"name": name}),
+            headers={
+                "X-Forwarded-User": "acct:{username}@{authority}".format(username=username, authority=authority),
+            },
+            timeout=1,
+        )
+        response.raise_for_status()
+    except requests.RequestException:
+        raise HTTPGatewayTimeout(explanation="Connecting to Hypothesis failed")
+
+    # Save a record of the group's pubid in the DB so that we can find it
+    # again later.
+    request.db.add(models.CourseGroup(
+        tool_consumer_instance_guid=tool_consumer_instance_guid,
+        context_id=context_id,
+        pubid=response.json()["id"],
```

Comment:
Do we need any error handling here?

Reply:
This would only crash if the h API returned successful response that did not have a JSON body or the JSON body contained no id. We could add error handling for it but realistically it doesn't happen, and if it did happen I think the behaviour of the app's current error handling might be to swallow the error so that we don't hear about it in Sentry! So I'd like to put this one off until later